libp2p-kad/src/query.rs

```diff
@@ -0,0 +1,335 @@
+    if bits_diff > 8 * (my_id_len - 2) {
+        return Err(());
+    }
+
+    let mut random_id = [0; 64];
+    for byte in 0..my_id_len {
+        match byte.cmp(&(my_id_len - bits_diff / 8 - 1)) {
+            Ordering::Less => {
+                random_id[byte] = my_id.as_bytes()[byte];
+            }
+            Ordering::Equal => {
+                let mask: u8 = (1 << (bits_diff % 8)) - 1;
+                random_id[byte] = (my_id.as_bytes()[byte] & !mask) | (rand::random::<u8>() & mask);
+            }
+            Ordering::Greater => {
+                random_id[byte] = rand::random();
+            }
+        }
+    }
+
+    let peer_id = PeerId::from_bytes(random_id[..my_id_len].to_owned())
+        .expect("randomly-generated peer ID should always be valid");
+    Ok(peer_id)
+}
+
+// Generic query-performing function.
+fn query<'a, I>(
+    query_interface: I,
+    searched_key: PeerId,
+    num_results: usize,
+) -> Box<Future<Item = Vec<PeerId>, Error = IoError> + 'a>
+where
+    I: QueryInterface + 'a,
+{
+    // State of the current iterative process.
+    struct State<'a> {
+        // If true, we are still in the first step of the algorithm where we try to find the
+        // closest node. If false, then we are contacting the k closest nodes in order to fill the
+        // list with enough results.
+        looking_for_closer: bool,
+        // Final output of the iteration.
+        result: Vec<PeerId>,
+        // For each open connection, a future with the response of the remote.
+        // Note that don't use a `SmallVec` here because `select_all` produces a `Vec`.
+        current_attempts_fut: Vec<Box<Future<Item = Vec<protocol::Peer>, Error = IoError> + 'a>>,
+        // For each open connection, the peer ID that we are connected to.
+        // Must always have the same length as `current_attempts_fut`.
+        current_attempts_addrs: SmallVec<[PeerId; 32]>,
+        // Nodes that need to be attempted.
+        pending_nodes: Vec<PeerId>,
+        // Peers that we tried to contact but failed.
+        failed_to_contact: FnvHashSet<PeerId>,
+    }
+
+    let initial_state = State {
+        looking_for_closer: true,
+        result: Vec::with_capacity(num_results),
+        current_attempts_fut: Vec::new(),
+        current_attempts_addrs: SmallVec::new(),
+        pending_nodes: query_interface.kbuckets_find_closest(&searched_key),
+        failed_to_contact: Default::default(),
+    };
+
+    let parallelism = query_interface.parallelism();
+
+    let stream = future::loop_fn(initial_state, move |mut state| {
+        let searched_key = searched_key.clone();
+        let query_interface = query_interface.clone();
+        let query_interface2 = query_interface.clone();
+
+        // Find out which nodes to contact at this iteration.
+        let to_contact = {
+            let mut to_contact = SmallVec::<[_; 16]>::new();
+            let wanted_len = if state.looking_for_closer {
```

Comment:
Comment this explaining what's going on. In general this function should be better-commented because it's huge and ugly and inpenetrable. I don't hate big functions or cyclomatic complexity but I think that they're a sign that you probably need good comments.